packages/sdk/android/src/main/java/com/primerioreactnative/components/manager/ach/PrimerRNHeadlessUniversalCheckoutStripeAchUserDetailsComponent.kt

```diff
@@ -0,0 +1,310 @@
+      promise.reject(exception.errorId, exception.description)
+    } else {
+      component?.updateCollectedData(AchUserDetailsCollectableData.FirstName(firstName))
+      promise.resolve(null)
+    }
+  }
+
+  @ReactMethod
+  fun onSetLastName(lastName: String, promise: Promise) {
+    if (component == null) {
+      val exception = ErrorTypeRN.NativeBridgeFailed errorTo UNINITIALIZED_ERROR
+      promise.reject(exception.errorId, exception.description)
+    } else {
+      component?.updateCollectedData(AchUserDetailsCollectableData.LastName(lastName))
+      promise.resolve(null)
+    }
+  }
+
+  @ReactMethod
+  fun onSetEmailAddress(emailAddress: String, promise: Promise) {
+    if (component == null) {
+      val exception = ErrorTypeRN.NativeBridgeFailed errorTo UNINITIALIZED_ERROR
+      promise.reject(exception.errorId, exception.description)
+    } else {
+      component?.updateCollectedData(AchUserDetailsCollectableData.EmailAddress(emailAddress))
+      promise.resolve(null)
+    }
+  }
+
+  private suspend fun configureErrorListener() {
+    component?.componentError?.collectLatest { error ->
+      sendEvent(
+          PrimerHeadlessUniversalCheckoutComponentEvent.ON_ERROR.eventName,
+          JSONObject().apply {
+            put(
+                "errors",
+                JSONArray().apply { put(JSONObject(json.encodeToString(error.toPrimerErrorRN()))) }
+            )
+          }
+      )
+    }
+  }
+
+  private suspend fun configureStepListener() {
+    component?.componentStep?.collectLatest { step ->
+      when (step) {
+        is AchUserDetailsStep.UserDetailsRetrieved -> {
+          sendEvent(
+              PrimerHeadlessUniversalCheckoutComponentEvent.ON_STEP.eventName,
+              JSONObject(json.encodeToString(step.toUserDetailsRetrievedRN()))
+          )
+        }
+        is AchUserDetailsStep.UserDetailsCollected -> {
+          sendEvent(
+              PrimerHeadlessUniversalCheckoutComponentEvent.ON_STEP.eventName,
+              JSONObject(json.encodeToString(step.toUserDetailsCollectedRN()))
+          )
+        }
+      }
+    }
+  }
+
+  private suspend fun configureValidationListener() {
+    component?.componentValidationStatus?.collectLatest { validationStatus ->
+      when (validationStatus) {
+        is PrimerValidationStatus.Validating -> {
+          sendEvent(
+              PrimerHeadlessUniversalCheckoutComponentEvent.ON_VALIDATING.eventName,
+              JSONObject().apply {
+                putData(validationStatus.collectableData as AchUserDetailsCollectableData)
+              }
+          )
+        }
+        is PrimerValidationStatus.Invalid -> {
+          sendEvent(
+              PrimerHeadlessUniversalCheckoutComponentEvent.ON_IN_VALID.eventName,
+              JSONObject().apply {
+                putData(validationStatus.collectableData as AchUserDetailsCollectableData)
+                put(
+                    "errors",
```

Comment:
[nit]: we seem to have this in every component, maybe worth centralising this and use in all components?

Reply:
Will do.